rust/lance/src/dataset.rs

```diff
@@ -823,77 +823,141 @@ impl Dataset {
+                })?;
+                let row_ids: Vec<u32> = row_ids[range].iter().map(|x| *x as u32).collect();
+                let batch = fragment.take_rows(&row_ids, projection, false).await?;
+                batches.push(batch);
+            }
+            Ok(concat_batches(&batches[0].schema(), &batches)?)
+        } else {
+            let projection_with_row_id = Schema::merge(
+                projection,
+                &ArrowSchema::new(vec![ArrowField::new(
+                    ROW_ID,
+                    arrow::datatypes::DataType::UInt64,
+                    false,
+                )]),
+            )?;
+            let schema_with_row_id = Arc::new(ArrowSchema::from(&projection_with_row_id));
+
+            // Slow case: need to re-map data into expected order
+            let mut sorted_row_ids = Vec::from(row_ids);
+            sorted_row_ids.sort();
+            // Group ROW Ids by the fragment
+            let mut row_ids_per_fragment: BTreeMap<u64, Vec<u32>> = BTreeMap::new();
+            sorted_row_ids.iter().for_each(|row_id| {
+                let fragment_id = row_id >> 32;
+                let offset = (row_id - (fragment_id << 32)) as u32;
+                row_ids_per_fragment
+                    .entry(fragment_id)
+                    .and_modify(|v| v.push(offset))
+                    .or_insert_with(|| vec![offset]);
+            });
 
-        // Remove the row id column.
-        let keep_indices = (0..one_batch.num_columns() - 1).collect::<Vec<_>>();
-        let one_batch = one_batch.project(&keep_indices)?;
-        let struct_arr: StructArray = one_batch.into();
-        let reordered = take(&struct_arr, &remapping_index, None)?;
-        Ok(as_struct_array(&reordered).into())
+            let fragments = self.get_fragments();
+            let fragment_and_indices = fragments.iter().filter_map(|f| {
+                let local_row_ids = row_ids_per_fragment.get(&(f.id() as u64))?;
+                Some((f, local_row_ids))
+            });
+
+            let mut batches = stream::iter(fragment_and_indices)
+                .then(|(fragment, indices)| fragment.take_rows(indices, projection, true))
+                .try_collect::<Vec<_>>()
+                .await?;
+
+            let one_batch = if batches.len() > 1 {
+                concat_batches(&schema_with_row_id, &batches)?
+            } else {
+                batches.pop().unwrap()
+            };
+            // Note: one_batch may contains fewer rows than the number of requested
+            // row ids because some rows may have been deleted. Because of this, we
+            // get the results with row ids so that we can re-order the results
+            // to match the requested order.
+
+            let returned_row_ids = one_batch
+                .column_by_name(ROW_ID)
+                .ok_or_else(|| Error::Internal {
+                    message: "ROW_ID column not found".into(),
+                })?
+                .as_any()
```

Comment:
is this written by copilot? we can just do `one_batch.column_by_name(...).ok_or_else().as_primiritve::<UInt64Type>`?